config/config-mysql_apache_jvm.yaml

```diff
@@ -0,0 +1,307 @@
+                new_value: sending
+              - value: startup
+                new_value: starting
+              - value: graceful_stop
+                new_value: finishing
+
+      - metric_name: apache_workers
+        action: update
+        new_name: agent.googleapis.com/apache/idle_workers
+        operations:
+          - action: aggregate_labels
+            label_set: []
+            aggregation_type: max
+            #this is a hack since the correct value is not necessarily max
+      
+      - metric_name: apache_sent_kilobytes_total
+        action: update
+        new_name: agent.googleapis.com/apache/traffic
+        operations:
+          - action: toggle_scalar_data_type
+
+  metricstransform/host:
+    transforms:
+      # system.cpu.time -> cpu/usage_time
+      - metric_name: system.cpu.time
+        action: update
+        new_name: agent.googleapis.com/cpu/usage_time
+        operations:
+          # change data type from double -> int64
+          - action: toggle_scalar_data_type
+          # change label cpu -> cpu_number
+          - action: update_label
+            label: cpu
+            new_label: cpu_number
+          # change label state -> cpu_state
+          - action: update_label
+            label: state
+            new_label: cpu_state
+          # take mean over cpu_number dimension, retaining only cpu_state
+          - action: aggregate_labels
+            label_set: [ cpu_state ]
+            aggregation_type: mean
+            
+      - metric_name: system.memory.usage
+        action: update
+        new_name: agent.googleapis.com/memory/bytes_used
+        operations:
+          # change data type from int64 -> double
+          - action: toggle_scalar_data_type
+          # aggregate state label values: slab_reclaimable & slab_unreclaimable -> slab (note this is not currently supported)
+          - action: aggregate_label_values
+            label: state
+            aggregated_values: [slab_reclaimable, slab_unreclaimable]
+            new_value: slab
+            aggregation_type: sum
+            
+      - metric_name: system.network.tcp_connections 
+        action: update
+        new_name: agent.googleapis.com/network/tcp_connections
+        operations:
+          # change data type from int64 -> double
+          - action: toggle_scalar_data_type
+          # change label state -> tcp_state
+          - action: update_label
+            label: state
+            new_label: tcp_state
+            
+      - metric_name: system.swap.paging_ops
+        action: update
+        new_name: agent.googleapis.com/swap/io
+        operations:
+          # delete singular type dimension, retaining only direction
+          - action: aggregate_labels
+            label_set: [ direction ]
+            aggregation_type: sum
+
+exporters:
+  stackdriver:  
+    project:  # stackdriver will use the default GCE credentials
+    metric_prefix: # empty to prevent the custom.googleapis.com/ prefix
```

Comment:
Don't we need to add `custom.googleapis.com`? If we don't wouldn't the metrics get rejected?

What's the intention here btw? Any explicitly listed transformed metrics get sent to the corresponding agent metrics, and any other metrics are exported as custom metrics?

Reply:
The idea here is that this config "mimicks" the agent and actually exports metrics in the same format as it would for third-party applications. So all the metrics in this file need to be prefixed by `agent.googleapis.com`. The reason why I hadn't just put the prefix in the `stackdriver` exporter was to allow some metrics to go through without that prefix, but I know now that you need a certain prefix for cloud moniroting. This means that within a certain receiver's metrics (say Java), I would need to filter out the "agent" metrics and put them in a separate pipeline from the "normal" ones. However, after discussions with Quentin, I will just whitelist the metrics wanted and ignore the rest, which means I can simply put them through the same exporter (I'll make an additional exporter which won't be used in this config but can be used by a user for custom metrics), and it will reduce the cost of additional metrics the user may not want. To recap I'll filter everything out but the metrics that are used to populate the pre-made dashboards on Cloud Monitoring which trigger automatically from the third-party metrics (JVM, MySQL and Apache)

Reply:
I think the intent to use the agent prefix for curated metrics and fall back to using the custom prefix for any other metrics is valid. To me it seems a bit limiting for users if they can't easily export any metrics that aren't on the list of curated agent metrics. I'm not sure that what you had in the config before was correct though, as I think you would need to explicitly add the custom prefix for any of the metrics that you don't have a transform set up - is that what happens when you set `metric_prefix` to blank? (I thought that would mean you don't set any prefix for non transformed metrics so they will end up not getting sent at all?).

But in terms of if it is important to let users export non-curated metrics, I'm not sure - Quentin would be more of an expert than me, so if he recommends doing it this way for now, then this LGTM.

Reply:
Yeah I agree it's limiting, so what I'll do is a setup similar to the one I had before. I'll manually set those "curated" metrics to have the `agent.googleapis.com`, and it'll be up to the user to add their own prefix when renaming if they want more than the "curated" metrics. At least this way they're not stuck with the `agent.googleapis.com` prefix on all the metrics which would make cloud monitoring block the extra ones, if they're not those curated ones. Doesn't sound ideal either way, but at least this way the user CAN get those metrics, they just need to transform their name to include a prefix. It comes back to the same for the agent metrics but limits the users less for the additional metrics. Do you agree?

Reply:
I think you should:

- For all agent metrics, set the prefix specifically as part of the transform
- In the stackdriver exporter, specify the custom prefix, so that any non-transformed metrics will use that prefix by default
- Ensure that in the default config you are filtering out all metrics that are not explicitly transformed to agent metrics

That way all the user needs to do if they want to include additional custom metrics is to change the filter (they can transform the metrics too if desired, but they shouldn't have to do that).

Reply:
I'm not sure about the second point I commented above actually, it's very hard to tell if that works as I expected from looking at the OpenCensus Go Exporter code. It would be nice if it did. Can you check? If not, then what you have is probably the best we can do for now.

Reply:
Okay! sounds good, could you check Gitter, I sent you a message there! Thanks!

Reply:
Replied.

As for my comment above, actually it should work as I stated as per: https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/pull/273

I'm not sure if that was every released though. I'm going to release this now and then create a PR to update contrib. After that, maybe we can try updating.

Reply:
Sent you another Gitter message (can't use Gchat sorry)

Reply:
Contrib was released with this change I believe. You can try updating the version of Contrib and testing if this now works (may help resolve your other dependency issue as well)

Reply:
updated the config, but still getting the same build error. If you pull this branch, do you get the same thing?